### PR TITLE
CORE-8433 Public API for filtering transactions

### PR DESCRIPTION
**Pull request by lankydan**

### CODE REVIEW

File: components/ledger/ledger-common-flow-api/src/main/kotlin/net/corda/ledger/common/flow/transaction/filtered/factory/ComponentGroupFilterParameters.kt [hunk 1]
```diff
@@ -28,7 +29,7 @@ sealed interface ComponentGroupFilterParameters {
      * @property componentGroupIndex The index of the component group to include.
      * @property deserializedClass The type that the component group deserializes its components into.
      */
-    data class AuditProof(override val componentGroupIndex: Int, val deserializedClass: Class<*>) : ComponentGroupFilterParameters {
+    data class AuditProof<T : Any>(override val componentGroupIndex: Int, val deserializedClass: Class<T>, val predicate: Predicate<T>) : ComponentGroupFilterParameters {
```
Comment (vlajos) [inline components/ledger/ledger-common-flow-api/src/main/kotlin/net/corda/ledger/common/flow/transaction/filtered/factory/ComponentGroupFilterParameters.kt]: docblock became outdated.

File: testing/ledger/ledger-utxo-testkit/src/main/kotlin/net/corda/ledger/utxo/testkit/UtxoSignedTransactionExample.kt [hunk 2]
```diff
@@ -13,13 +14,15 @@ import net.corda.v5.application.crypto.DigestService
 import net.corda.v5.application.marshalling.JsonMarshallingService
 import net.corda.v5.application.serialization.SerializationService
 import net.corda.v5.ledger.utxo.transaction.UtxoSignedTransaction
+import java.time.Instant
```
Comment (vlajos) [inline testing/ledger/ledger-utxo-testkit/src/main/kotlin/net/corda/ledger/utxo/testkit/UtxoSignedTransactionExample.kt]: Is this used?

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/filtered/UtxoFilteredTransactionBuilderImpl.kt [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    override val signatories: ComponentGroupFilterParameters? = null,
+    override val inputStates: ComponentGroupFilterParameters? = null,
+    override val referenceInputStates: ComponentGroupFilterParameters? = null,
+    override val outputStates: ComponentGroupFilterParameters? = null,
+    override val commands: ComponentGroupFilterParameters? = null
+) : UtxoFilteredTransactionBuilder, UtxoFilteredTransactionBuilderInternal {
+
+    @Suspendable
+    override fun withNotary(): UtxoFilteredTransactionBuilderInternal {
+        // only filters out the notary, adjust the lambda to also filter out time window
```
Comment (vlajos) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/filtered/UtxoFilteredTransactionBuilderImpl.kt]: is this a todo?

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/filtered/UtxoFilteredTransactionBuilderImpl.kt [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            )
+        )
+    }
+
+    @Suspendable
+    override fun withOutputStatesSize(): UtxoFilteredTransactionBuilderInternal {
+        return copy(outputStates = ComponentGroupFilterParameters.SizeProof(UtxoComponentGroup.OUTPUTS.ordinal))
+    }
+
+    @Suspendable
```
Comment (blsemo) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/filtered/UtxoFilteredTransactionBuilderImpl.kt:107]: If you filter the output states, you need to filter the output state infos as well. That's a bit of a pain with the predicates, of course.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/filtered/UtxoFilteredTransactionBuilderImpl.kt [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        )
+    }
+
+    @Suspendable
+    override fun withCommandsSize(): UtxoFilteredTransactionBuilderInternal {
+        return copy(commands = ComponentGroupFilterParameters.SizeProof(UtxoComponentGroup.COMMANDS.ordinal))
+    }
+
+    @Suspendable
+    override fun withCommands(): UtxoFilteredTransactionBuilderInternal {
```
Comment (blsemo) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/filtered/UtxoFilteredTransactionBuilderImpl.kt:129]: Similar to outputs - if you filter the commands, you want to apply the same filter to the command infos.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/filtered/UtxoFilteredTransactionBuilderImpl.kt (hunk 3 above)
Comment (blsemo) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/filtered/UtxoFilteredTransactionBuilderImpl.kt]: That filter doesn't look right - I think we want to be able to specify to filter either notary or timeWindow or both. See API PR.

**Review by blsemo — DISMISSED** [review]
The rules for filtering the outputs need to be fixed to match what the `UtxoFilteredTransactionImpl` expects to find in the filtered transaction.
Ideally write a test that use the UtxoFiltereredTransactionBuilder and then tries to load filtered outputs.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/filtered/factory/UtxoFilteredTransactionFactoryImpl.kt [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    @Reference(service = SerializationService::class)
+    private val serializationService: SerializationService
+) : UtxoFilteredTransactionFactory, UsedByFlow {
+
+    @Suspendable
+    override fun create(
+        signedTransaction: UtxoSignedTransactionInternal,
+        filteredTransactionBuilder: UtxoFilteredTransactionBuilderInternal
+    ): UtxoFilteredTransaction {
+        val notaryAndTimeWindow = if (filteredTransactionBuilder.notary || filteredTransactionBuilder.timeWindow) {
```
Comment (blsemo) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/filtered/factory/UtxoFilteredTransactionFactoryImpl.kt:40]: Nice, that's the kind of logic I was thinking of.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/filtered/factory/UtxoFilteredTransactionFactoryImpl.kt [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            serializationService,
+            filteredTransactionFactory.create(
+                signedTransaction.wireTransaction,
+                listOfNotNull(
+                    ComponentGroupFilterParameters.AuditProof(METADATA.ordinal, TransactionMetadata::class.java) { true },
+                    notaryAndTimeWindow,
+                    filteredTransactionBuilder.signatories,
+                    filteredTransactionBuilder.inputStates,
+                    filteredTransactionBuilder.referenceInputStates,
+                    filteredTransactionBuilder.outputStates?.let { _ ->
```
Comment (blsemo) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/filtered/factory/UtxoFilteredTransactionFactoryImpl.kt]: You probably need to check if the outputStates are an AUDIT proof - if there's a size only proof, we shouldn't include the infos. Same for commands.

File: components/ledger/ledger-utxo-flow/src/test/kotlin/net/corda/ledger/utxo/flow/impl/transaction/filtered/factory/UtxoFilteredTransactionFactoryImplTest.kt [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        assertThat(componentGroupFilterParameters).hasSize(3)
+        assertThat(componentGroups).containsExactly(
+            UtxoComponentGroup.METADATA.ordinal,
+            UtxoComponentGroup.OUTPUTS_INFO.ordinal,
+            UtxoComponentGroup.OUTPUTS.ordinal
+        )
+    }
+
+    @Test
+    fun `does not include output state infos if the output state component group is not included`() {
```
Comment (blsemo) [inline components/ledger/ledger-utxo-flow/src/test/kotlin/net/corda/ledger/utxo/flow/impl/transaction/filtered/factory/UtxoFilteredTransactionFactoryImplTest.kt:144]: Should have a test `does not inlude output state infos if the output state component group is size only proof`.

**Review by blsemo — COMMENTED** [review]
Nearly there, but I think there is still a slight logic flaw as in that the output state infos are inluded when output states are a size only proof (and that is a very common use case for the notary, so we need to make sure it's correct and tested properly).

**Review by blsemo — DISMISSED** [review]
LGTM

**Review by blsemo — APPROVED** [review]

**Pull request merged**